widgets/heatmap/heatmap.go

```diff
@@ -89,42 +144,161 @@ func (hp *HeatMap) ClearYLabels() {
 	hp.yLabels = nil
 }
 
-// ValueCapacity returns the number of values that can fit into the canvas.
+// ValueCapacity returns the number of rows and columns of the heat map
+// that can fit into the canvas.
+//
 // This is essentially the number of available cells on the canvas as observed
 // on the last call to draw. Returns zero if draw wasn't called.
 //
 // Note that this capacity changes each time the terminal resizes, so there is
 // no guarantee this remains the same next time Draw is called.
 // Should be used as a hint only.
-func (hp *HeatMap) ValueCapacity() int {
-	return 0
+func (hp *HeatMap) ValueCapacity() (rows, cols int) {
+	hp.mu.Lock()
+	defer hp.mu.Unlock()
+
+	if hp.lastWidth == 0 || hp.lastHeight == 0 {
+		return 0, 0
+	}
+
+	rows = hp.lastHeight - 1
+	var cw int
+
+	if hp.opts.cellWidth > minCellWidth {
+		cw = hp.opts.cellWidth
+	} else {
+		cw = minCellWidth
+	}
+	cols = int(math.Floor(float64(hp.lastWidth-axes.LongestString(hp.yLabels)-axes.AxisWidth) / float64(cw)))
+	return
 }
 
 // axesDetails determines the details about the X and Y axes.
 func (hp *HeatMap) axesDetails(cvs *canvas.Canvas) (*axes.XDetails, *axes.YDetails, error) {
-	return nil, nil, errors.New("not implemented")
+	hp.cellWidthAdaptive(cvs)
+
+	yd, err := axes.NewYDetails(hp.yLabels)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	xd, err := axes.NewXDetails(yd.End, hp.xLabels, hp.opts.cellWidth)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return xd, yd, nil
 }
 
 // Draw draws cells, X labels and Y labels as HeatMap.
 // Implements widgetapi.Widget.Draw.
 func (hp *HeatMap) Draw(cvs *canvas.Canvas, meta *widgetapi.Meta) error {
-	return errors.New("not implemented")
+	hp.mu.Lock()
```

Comment:
Thank you for continuing to work on this. On top of the bug that you identified, we also seem to have one when the widget's window is resized down to just 7 characters of width:

```
panic: runtime error: integer divide by zero

goroutine 1 [running]:
github.com/mum4k/termdash/widgets/heatmap.(*HeatMap).cellWidthAdaptive(0xc000166120, 0xc0002049c0)
	/Users/mum4k/go/src/github.com/mum4k/termdash/widgets/heatmap/heatmap.go:275 +0xb8
github.com/mum4k/termdash/widgets/heatmap.(*HeatMap).axesDetails(0xc000166120, 0xc0002049c0, 0x0, 0x0, 0x3, 0x19)
	/Users/mum4k/go/src/github.com/mum4k/termdash/widgets/heatmap/heatmap.go:178 +0x39
github.com/mum4k/termdash/widgets/heatmap.(*HeatMap).Draw(0xc000166120, 0xc0002049c0, 0xc0004fc810, 0x0, 0x0)
	/Users/mum4k/go/src/github.com/mum4k/termdash/widgets/heatmap/heatmap.go:210 +0x14c
github.com/mum4k/termdash/container.drawWidget(0xc00013e700, 0x0, 0x0)
	/Users/mum4k/go/src/github.com/mum4k/termdash/container/draw.go:138 +0x2bb
github.com/mum4k/termdash/container.drawCont(0xc00013e700, 0x1, 0x1)
	/Users/mum4k/go/src/github.com/mum4k/termdash/container/draw.go:171 +0x147
github.com/mum4k/termdash/container.drawTree.func1(0xc00013e700, 0x7, 0x1b)
	/Users/mum4k/go/src/github.com/mum4k/termdash/container/draw.go:63 +0x85
github.com/mum4k/termdash/container.preOrder(0xc00013e700, 0xc000265bd8, 0x119d550)
	/Users/mum4k/go/src/github.com/mum4k/termdash/container/traversal.go:43 +0x4f
github.com/mum4k/termdash/container.drawTree(0xc00013e700, 0x1b, 0x0)
	/Users/mum4k/go/src/github.com/mum4k/termdash/container/draw.go:43 +0x125
github.com/mum4k/termdash/container.(*Container).Draw(0xc00013e700, 0x0, 0x0)
	/Users/mum4k/go/src/github.com/mum4k/termdash/container/container.go:233 +0x11c
github.com/mum4k/termdash.(*termdash).redraw(0xc000134420, 0x0, 0x0)
	/Users/mum4k/go/src/github.com/mum4k/termdash/termdash.go:274 +0x3d
github.com/mum4k/termdash.(*termdash).periodicRedraw(0xc000134420, 0x0, 0x0)
	/Users/mum4k/go/src/github.com/mum4k/termdash/termdash.go:301 +0x71
github.com/mum4k/termdash.(*termdash).start(0xc000134420, 0x11c4f40, 0xc000137100, 0x0, 0x0)
	/Users/mum4k/go/src/github.com/mum4k/termdash/termdash.go:344 +0x216
```